presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java

```diff
@@ -339,6 +359,49 @@ public ConnectorPageSource getPageSource(
         }
     }
 
+    Optional<BitSet> getRowsFromUuid(FileSystem fileSystem, Optional<UUID> deltaShardUuid)
+    {
+        if (!deltaShardUuid.isPresent()) {
+            return Optional.empty();
+        }
+
+        try (OrcDataSource dataSource = openShard(fileSystem, deltaShardUuid.get(), defaultReaderAttributes)) {
+            AggregatedMemoryContext systemMemoryUsage = newSimpleAggregatedMemoryContext();
+            OrcReader reader = new OrcReader(
+                    dataSource,
+                    ORC,
+                    orcFileTailSource,
+                    new StorageStripeMetadataSource(),
+                    new OrcReaderOptions(
+                            defaultReaderAttributes.getMaxMergeDistance(),
+                            defaultReaderAttributes.getTinyStripeThreshold(),
+                            HUGE_MAX_READ_BLOCK_SIZE,
+                            defaultReaderAttributes.isZstdJniDecompressionEnabled()));
+
+            if (reader.getFooter().getNumberOfRows() >= Integer.MAX_VALUE) {
+                throw new IOException("File has too many rows");
+            }
+
+            OrcBatchRecordReader recordReader = reader.createBatchRecordReader(
```

Comment:
> a `try` is needed to avoid file descriptor leak. Check other use cases.

I think as long as we close ``OrcDataSource dataSource``,  we wouldn't have file descriptor leak.

So in above, we use 
``try (OrcDataSource dataSource = openShard(fileSystem, deltaShardUuid.get(), defaultReaderAttributes)) ``
which will close the datasource once it encounters some exceptions

Reply:
You will miss closing all the stream readers. So use `try` to make sure it is closed properly.

Reply:
> You will miss closing all the stream readers. So use `try` to make sure it is closed properly.

But I do see from other examples that they used OrcBatchRecordReader with ``OrcDataSource dataSource`` as well, and they only close ``dataSource``
eg1:
OrcStorageManager
public ConnectorPageSource getPageSource
it uses try and catch and it only closes the dataSource``closeQuietly(dataSource);``

eg2:
OrcBatchPageSourceFactory
createOrcPageSource
``orcDataSource.close();``

Reply:
Because readers will be closed by `ConnectorPageSource`. In your case, no one can close the readers.